### ISSUE TITLE
Docs need to explain about Minio client class type change introduced in 6.0 release

### ISSUE THREAD

**Issue by ollie10** [minio/minio-dotnet#891]
Hi Everyone!

It seems that **WithRetryPolicy** method was broken after the update to v 6.0

I can't find any working version so far not in the documentation neither in GitHub.

I had this code working before that now cannot compile because the WithRetryPolicy extension method changed.

This in my class

```csharp
public async Task UploadImage(string fileName, string fileSystemFile)
{
	MinioClient client = new MinioClient().WithEndpoint(_appSettings.MediaConfiguration.BucketEndpoint).WithCredentials(_appSettings.MediaConfiguration.BucketAccessKey, _appSettings.MediaConfiguration.BucketSecretKey).WithRegion(_appSettings.MediaConfiguration.BucketRegion).WithSSL().WithTimeout(30000).WithRetryPolicy(RetryPolicyHelper.GetDefaultRetryPolicy(3, TimeSpan.FromSeconds(2), TimeSpan.FromSeconds(30))).Build();

	PutObjectArgs arguments = new PutObjectArgs().WithBucket(_appSettings.MediaConfiguration.BucketName).WithObject(fileName).WithFileName(fileSystemFile).WithContentType("image/jpeg");
	await client.PutObjectAsync(arguments);
}
```

This was the helper:

```csharp
internal static class RetryPolicyHelper
{
	private const int defaultRetryCount = 3;
	private static readonly TimeSpan defaultRetryInterval = TimeSpan.FromSeconds(30);
	private static readonly TimeSpan defaultMaxRetryInterval = TimeSpan.FromSeconds(300);
	// exponential backoff with jitter, truncated by max retry interval
	public static TimeSpan CalcBackoff(int retryAttempt, TimeSpan retryInterval, TimeSpan maxRetryInterval)
	{
		// 0.8..1.2
		var jitter = 0.8 + new Random(Environment.TickCount).NextDouble() * 0.4;
		// (2^retryCount - 1) * jitter
		var scaleCoeff = (Math.Pow(2.0, retryAttempt) - 1.0) * jitter;
		// Apply scale coefficient
		var result = TimeSpan.FromMilliseconds(retryInterval.TotalMilliseconds * scaleCoeff);
		// Truncate by max retry interval
		return result < maxRetryInterval ? result : maxRetryInterval;
	}

	public static PolicyBuilder<ResponseResult> CreatePolicyBuilder()
	{
		return Policy<ResponseResult>.Handle<ConnectionException>().Or<InternalClientException>(ex => ex.Message.StartsWith("Unsuccessful response from server"));
	}

	public static AsyncPolicy<ResponseResult> GetDefaultRetryPolicy() => GetDefaultRetryPolicy(defaultRetryCount, defaultRetryInterval, defaultMaxRetryInterval);

	public static AsyncPolicy<ResponseResult> GetDefaultRetryPolicy(int retryCount, TimeSpan retryInterval, TimeSpan maxRetryInterval) => CreatePolicyBuilder().WaitAndRetryAsync(retryCount, i => CalcBackoff(i, retryInterval, maxRetryInterval));

        // THIS LINE DOESN'T COMPILE
	public static RetryPolicyHandlingDelegate AsRetryDelegate(this AsyncPolicy<ResponseResult> policy) => policy == null ? (RetryPolicyHandlingDelegate)null : async executeCallback => await policy.ExecuteAsync(executeCallback);

	public static MinioClient WithRetryPolicy(this MinioClient client, AsyncPolicy<ResponseResult> policy) => client.WithRetryPolicy(policy.AsRetryDelegate());
}
```

Any working example of how should be the new Retry Policy Delegate?
Many thanks
Best regards

**Comment by loekensgard** [comment]
Is there a workaround for this issue? We have the same problem.

**Comment by ebozduman** [comment]
@loekensgard, @ollie10,

Sorry for the delay. 

I'll check my to do list and start looking into it as soon as possible. I'll add it in my high priority list of issues as this looks like a regression.

**Comment by ebozduman** [comment]
@loekensgard, @ollie10,

This is due to the change/PR #856 in the major release 6.0.

    public static MinioClient WithRetryPolicy(this **MinioClient** minioClient,
    ==>
    public static IMinioClient WithRetryPolicy(this **IMinioClient** minioClient,

So, the way to create a MinIO Client after this change is shown in an example in the `README.md` file, but I think the release notes should also have some explanation about it, although most of us would still miss it.

So, basically, you just need to change the way you create MinIO Client script, like:

`var client = new MinioClient().WithEndpoint(....` in your `UploadImage` method,
        and

`	public static IMinioClient WithRetryPolicy(this IMinioClient client, AsyncPolicy<ResponseResult> policy) => client.WithRetryPolicy(policy.AsRetryDelegate());` in your `RetryPolicyHelper` class definition.

Please let me know if the above explanations help.

**Comment by ebozduman** [comment]
@loekensgard, @ollie10,

Renaming the title as a documentation issue.
Please let me know if you disagree.

**Comment by ollie10** [comment]
@ebozduman many thanks, I will give it a try tomorrow and let you know if it works as expected, Thanks again!

**Comment by loekensgard** [comment]
@ebozduman it works 👍 

Program.cs
```csharp
      builder.Services.AddMinio(config => config
           .WithEndpoint("Host")
           .WithCredentials("AccessKey", "SecreyKey")
           .WithRetryPolicy(new MinioRetryPolicy(2)));
```

MinioRetryPolicy.cs
```csharp
    public class MinioRetryPolicy(int retryCount) : IRetryPolicyHandler
    {
        private readonly IAsyncPolicy<ResponseResult> _retryPolicy = CreatePolicy(retryCount);

        private const int _retryIntervalInSeconds = 1;
        private const string _unsuccessfulResponseMessage = "Unsuccessful response from server";

        private static IAsyncPolicy<ResponseResult> CreatePolicy(int retryCount) => Policy<ResponseResult>
                .Handle<ConnectionException>()
                .Or<InternalClientException>(ex => ex.Message.StartsWith(_unsuccessfulResponseMessage))
                .WaitAndRetryAsync(Backoff.DecorrelatedJitterBackoffV2(TimeSpan.FromSeconds(_retryIntervalInSeconds), retryCount));

        public async Task<ResponseResult> Handle(Func<Task<ResponseResult>> executeRequestCallback)
        {
            return await _retryPolicy.ExecuteAsync(executeRequestCallback);
        }
    }
```
